configure.ps1

```diff
@@ -55,75 +57,17 @@ Invoke-BuildStep 'Cleaning package cache' {
 $ConfigureObject = @{
     BuildTools = @{}
     Toolsets = @{}
-}
-
-Function New-BuildToolset {
-    param(
-        [ValidateSet(15, 16)]
-        [int]$ToolsetVersion
-    )
-    $CommonToolsVar = "Env:VS${ToolsetVersion}0COMNTOOLS"
-    if (Test-Path $CommonToolsVar) {
-        $CommonToolsValue = gci $CommonToolsVar | select -expand value -ea Ignore
-        Verbose-Log "Using environment variable `"$CommonToolsVar`" = `"$CommonToolsValue`""
-        $ToolsetObject = @{
-            VisualStudioInstallDir = [System.IO.Path]::GetFullPath((Join-Path $CommonToolsValue '..\IDE'))
-        }
-    }
-
-    if (-not $ToolsetObject) {
-        $VisualStudioRegistryKey = "HKCU:\SOFTWARE\Microsoft\VisualStudio\${ToolsetVersion}.0_Config"
-        if (Test-Path $VisualStudioRegistryKey) {
-            Verbose-Log "Retrieving Visual Studio installation path from registry '$VisualStudioRegistryKey'"
-            $ToolsetObject = @{
-                VisualStudioInstallDir = gp $VisualStudioRegistryKey | select -expand InstallDir -ea Ignore
-            }
-        }
-    }
-
-    if (-not $ToolsetObject) {
-        $VisualStudioInstallRootDir = Get-LatestVisualStudioRoot
-
-        if ($VisualStudioInstallRootDir) {
-            Verbose-Log "Using willow instance '$VisualStudioInstallRootDir' installation path"
-            $ToolsetObject = @{
-                VisualStudioInstallDir = [System.IO.Path]::GetFullPath((Join-Path $VisualStudioInstallRootDir Common7\IDE\))
-            }
-        }
-    }
-
-    if (-not $ToolsetObject) {
-        $DefaultInstallDir = Join-Path $env:ProgramFiles "Microsoft Visual Studio ${ToolsetVersion}.0\Common7\IDE\"
-        if (Test-Path $DefaultInstallDir) {
-            Verbose-Log "Using default location of Visual Studio installation path"
-            $ToolsetObject = @{
-                VisualStudioInstallDir = $DefaultInstallDir
-            }
-        }
-    }
-
-    if (-not $ToolsetObject) {
-        Warning-Log "Toolset VS${ToolsetVersion} is not found."
-    }
-
-    # return toolset build configuration object
-    $ToolsetObject
-}
-
-$ProgramFiles = ${env:ProgramFiles(x86)}
-
-if (-not $ProgramFiles -or -not (Test-Path $ProgramFiles)) {
-    $ProgramFiles = $env:ProgramFiles
+    EnvVars = @{}
 }
 
 $vsMajorVersion = Get-VSMajorVersion
 $validateToolsetMessage = "Validating VS $vsMajorVersion toolset installation" 
 
 Invoke-BuildStep $validateToolsetMessage {
-
     $vstoolset = New-BuildToolset $vsMajorVersion
     if ($vstoolset) {
         $ConfigureObject.Toolsets.Add('vstoolset', $vstoolset)
+        $ConfigureObject.EnvVars.Add('VisualStudioVersion', "$vsMajorVersion.0")
```

Comment:
Why do we need this?
I don't see it consumed anywhere.

Reply:
`VisualStudioVersion` envvar is used specially in our build\build.proj for restoring and building. Remember the `Restoring for VS 16.0 ...` message.

I need to set this variable according to the VS found in configure.ps1 script. This more relevant when you're running the script in a plain Powershell prompt.

Reply:
I'm not sure how this works. 
EnvVars is just a new property on the ConfigureObject. 
The ConfigureObject is really just a json rep of what to build with. 
Aren't you missing a change that expands this envvar list? 

* The developer command prompt *sets* the VisualStudioVersion environment variable, can this mess with that? 
* The toolset shouldn't decide what version of VS you are building for. The `Restoring for VS 16.0` message is about the target, not the toolset being used.

Reply:
> I'm not sure how this works.
> EnvVars is just a new property on the ConfigureObject.
> The ConfigureObject is really just a json rep of what to build with.

Yes, see answer below.

> Aren't you missing a change that expands this envvar list?

It's already there on common.ps1:

https://github.com/NuGet/NuGet.Client/blob/f298d7187b8ef797ac16c05cacbae1c8f6d6eaa1/build/common.ps1#L368-L371

> A few other questions:
> 
> * The developer command prompt _sets_ the VisualStudioVersion environment variable, can this mess with that?

Yes, I did that change because, for some reason, msbuild always use 16. See output below

```
Testing dev16 EnvVar VisualStudioVersion
  Current VS: 16.0
Testing dev17 EnvVar VisualStudioVersion
  Current VS: 16.0
```

> * The toolset shouldn't decide what version of VS you are building for. The `Restoring for VS 16.0` message is about the target, not the toolset being used.

Reply:
> Yes, I did that change because, for some reason, msbuild always use 16. See output below

The property is set in https://github.com/NuGet/NuGet.Client/blob/f298d7187b8ef797ac16c05cacbae1c8f6d6eaa1/build/common.project.props#L66.

Reply:
I took a deeper look, I think it's 100% possible for us to remove the `VisualStudioVersion` usage from our scripts after your change has been merged. 

It was only there because we used to build from the same repo both dev14 and dev15. 
That's no longer the case.

Reply:
Ok, I will remove the envvar. For now, build script will build dev16 from vanilla PowerShell. dev17 should use Developer Command Prompt/PowerShell.

Reply:
I don't think you need to remove it.

Based on what I saw in the scripts, it shouldn't break anything.

Reply:
Filed a bug in MSBuild https://github.com/dotnet/msbuild/issues/6360

Reply:
oh welp haha. 

I think it's something they'll fix eventually. 
It takes time to propagate versioning :) 

Resolving this thread.